### PR TITLE
[CASSANDRA-631] Add an Offer Queue for more reliable Offer Processing

### PR DESCRIPTION
**Pull request by gabrielhartmann**
**Problem**:
We implicitly had a queue for processing Offers.  As each bundle of Offers came in we processed that bundle on a single thread of an executor service.  If another offer bundle came in it was queued.  In a large cluster it's possible to receive offers faster than we can process them.

In a scale test on a large cluster we saw that we attempted to consume Offers which had been in the queue so long that they were rescinded by Mesos.  We tried to process them anyway resulting in problems, but that's a different topic (https://jira.mesosphere.com/browse/INFINITY-1964).

**Solution**:
Build an explicit bounded queue (buffer) for offers.  We bound the queue to 100 offers by default.  This isn't currently configurable at runtime.  Until the need arises I'd like to keep it this way.  Offers are pushed onto the queue synchronously if there is capacity.  Placement of Offers into the in memory queue should happen pretty much instantaneously, so a separate thread isn't needed.  Another thread pulls offers off the queue as fast as it can.  On each iteration, this thread pulls _ALL_ offers out of the queue, so it processes 100 offers at a time at maximum.

**Test**:
Deployment of a 100 node Cassandra cluster failed on node-52 after ~3-4 hours of deploying without this change.  A 9 hour deployment of all 100 nodes succeeded with this change.

Note:
Please ignore the log level related changes.  They're based on an as yet unrelated but unmerged PR.  It'll go in before this PR.  It's somewhat related because somethings have been moved to `debug` log level.  It was observed that logs were pretty hard to follow in a very large service (C* with 100 nodes).

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
     // Mesos may call registered() multiple times in the lifespan of a Scheduler process, specifically when there's
     // master re-election. Avoid performing initialization multiple times, which would cause resourcesQueue to be stuck.
     private final AtomicBoolean isAlreadyRegistered = new AtomicBoolean(false);
+    protected final OfferQueue offerQueue;
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Should offers be getting queued in `UninstallScheduler`? From its perspective, offerRescinded will attempt to remove offers that aren't being added. Would it make sense for `UninstallScheduler` to take advantage of this as well?

File: frameworks/cassandra/universe/config.json [hunk 2]
```diff
@@ -55,6 +55,12 @@
           "description": "Enable DC/OS Virtual Networking",
           "type": "boolean",
           "default": false
+        },
+        "log_level": {
+          "description": "The log level for the DC/OS service.",
```
Comment (nickbp) [inline frameworks/cassandra/universe/config.json]: This name and description imply that it'll carry through to the underlying service. Maybe go with `scheduler_log_level`/`The log level for the service scheduler` or similar

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 3]
```diff
@@ -61,7 +60,11 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultScheduler.class);
 
-    protected final ExecutorService executor = Executors.newFixedThreadPool(1);
+    /**
+     * 1 Thread for processing offers off the queue in {@link #processOffers()}
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: With this structure, both threads could be used by a single operation (e.g. two concurrent processOffers() calls taking up the whole pool). Would it make sense to have two isolated queues?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 4]
```diff
@@ -587,12 +600,22 @@ protected DefaultScheduler(
...
+        processOffers();
     }
 
     @VisibleForTesting
-    void awaitTermination() throws InterruptedException {
-        executor.shutdown();
-        executor.awaitTermination(AWAIT_TERMINATION_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    void awaitOffersProcessed() throws InterruptedException {
+        while (!offerQueue.isEmpty()) {
+            LOGGER.warn("Queue is NOT empty");
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: Could include `... sleeping for 1s` in these messages

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+/**
+ * This class acts as a buffer of Offers from Mesos.  By default it holds a maximum of 100 Offers.
+ */
+public class OfferQueue {
+    public static final int DEFAULT_CAPACITY = 100;
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final BlockingQueue<Protos.Offer> queue;
+
+    public OfferQueue () {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: extra space. `OfferQueue()`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.stream.Collectors;
+
+/**
+ * This class acts as a buffer of Offers from Mesos.  By default it holds a maximum of 100 Offers.
+ */
+public class OfferQueue {
+    public static final int DEFAULT_CAPACITY = 100;
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Could be package-private and `@VisibleForTesting`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * Calling this method is a blocking Operation. It returns all Offers currently in the queue. It always returns at
+     * least one Offer.
+     */
+    public List<Protos.Offer> takeAll() {
+        List<Protos.Offer> offers = new LinkedList<>();
+        try {
+            offers.add(queue.take());
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Should add a comment here describing the blocking behavior/why both calls are needed

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            queue.drainTo(offers);
+        } catch (InterruptedException e) {
+            logger.warn("Interrupted while waiting for offer in queue.");
+        }
+
+        return offers;
+    }
+
+    /**
+     * This method enqueues an Offer from Mesos if their is capacity.
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: `there`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    public void remove(Protos.OfferID offerID) {
+        Collection<Protos.Offer> offers = queue.parallelStream()
+                .filter(offer -> offer.getId().equals(offerID))
+                .collect(Collectors.toList());
+
+        boolean removed = queue.removeAll(offers);
+        if (!removed) {
+            logger.warn(
+                    String.format(
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Should be able to use `{}` style formatting here (no exception being passed)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        }
+
+        return offers;
+    }
+
+    /**
+     * This method enqueues an Offer from Mesos if their is capacity.
+     * @return true if the Offer was successfully put in the queue, false otherwise
+     */
+    public boolean offer(Protos.Offer offer) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: `put()`?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * This method returns the number of elements in the queue.
+     */
+    public int getSize() {
+        return queue.size();
+    }
+
+    /**
+     * This method returns the remaining capacity in the queue.
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: What happens if the capacity is exhausted? Blocking or dropped data? (Can't hurt to mention here)

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java (hunk 8 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Can't hurt to also describe what happens if there isn't capacity. Interface implies the record isn't added?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/ParentElement.java [hunk 12]
```diff
@@ -17,7 +18,7 @@
  * @param <C> the type of the child elements
  */
 public interface ParentElement<C extends Element> extends Element, Interruptible {
-
+    public static final Logger LOGGER = LoggerFactory.getLogger(ParentElement.class);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/ParentElement.java]: `private`?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    public int getRemainingCapacity() {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Given that this is only used for tests, it could be package-private + `@VisibleForTesting`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+     * This method specifies whether any offers are in the queue.
+     */
+    public boolean isEmpty() {
+        return queue.isEmpty();
+    }
+
+    /**
+     * This method returns the number of elements in the queue.
+     */
+    public int getSize() {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Given that this is only used for tests, it could be package-private + `@VisibleForTesting`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java [hunk 15]
```diff
@@ -587,12 +600,22 @@ protected DefaultScheduler(
...
+            Thread.sleep(1000);
+        }
+
+        while (!offersInProgress.isEmpty()) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: There is a window/race between grabbing the offers from offerQueue (L782) and adding them to offersInProgress (L788) in which both lists would be empty.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 1 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Yes, I remembered and then forgot again about the UninstallScheduler, thanks for reminding me.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java (hunk 3 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/DefaultScheduler.java]: The `processOffers` call never releases its thread and the ordering of Scheduler startup precludes this possibility right now, but we could have separate executor services to avoid future developers hitting this accidentally.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java (hunk 10 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: Borrowing words from here, where `offer` means non-blocking and `put` means blocking.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java (hunk 11 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: I would mention this over at `offer`.  Seems irrelevant to `getReminaingCapacity`'s behavior.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/ParentElement.java (hunk 12 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/ParentElement.java]: Best I can do in an interface is package-private apparently.

Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/plan/ParentElement.java]: Oh didn't realize it was an interface

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java (hunk 11 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: :+1:

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java (hunk 10 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/OfferQueue.java]: :+1:

**Comment by gabrielhartmann** [comment]
@nickbp:  Comments addressed.  Probably worth another look at `AbstractScheduler` `UninstallScheduler` and `DefaultScheduler` changes.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 16]
```diff
@@ -63,8 +83,73 @@ public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, P
...
 
+    protected abstract boolean apiServerReady();
+
+    /**
+     * All offers must have been presented to resourceOffers() before calling this.
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    public void awaitOffersProcessed() throws InterruptedException {
+        while (!offersInProgress.isEmpty()) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Once offersInProgress is lock protected, a Condition could be used to synchronously wait for offersInProgress to be empty: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/Condition.html

in `processOffers()`:
```
lock.lock();
...
offersInProgress.removeAll(...);
if (offersInProgress.isEmpty()) {
  offersAreEmpty.signalAll();
}
...
lock.unlock();
```

here:

```
lock.lock();
while (!offersInProgress.isEmpty()) {
  offersAreEmpty.await();
}
lock.unlock();
```

File: sdk/scheduler/src/test/java/com/mesosphere/sdk/scheduler/OfferQueueTest.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+public class OfferQueueTest {
+    @Test
+    public void testEmptyQueue() {
+        OfferQueue offerQueue = new OfferQueue();
+        Assert.assertTrue(offerQueue.isEmpty());
+    }
+
+    @Test
+    public void testEnqueueOffer() {
+        OfferQueue offerQueue = new OfferQueue();
```
Comment (nickbp) [inline sdk/scheduler/src/test/java/com/mesosphere/sdk/scheduler/OfferQueueTest.java]: Instead of requiring access to OfferQueue.DEFAULT_CAPACITY, these tests could define their own TEST_CAPACITY and pass that to the OfferQueue constructor. Then DEFAULT_CAPACITY could remain `private`

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 18]
```diff
@@ -23,14 +29,28 @@
...
     private final AtomicBoolean isAlreadyRegistered = new AtomicBoolean(false);
+    protected final OfferQueue offerQueue = new OfferQueue();
     protected SchedulerDriver driver;
     protected DefaultReconciler reconciler;
+    private Set<Protos.OfferID> offersInProgress = new HashSet<>();
+
+    /**
+     * Executor for handling TaskStatus updates in {@link #statusUpdate(SchedulerDriver, Protos.TaskStatus)}.
+     */
+    protected final ExecutorService statusExecutor = Executors.newSingleThreadExecutor();
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: I just realized that with this change we're moving to status handling and offer handling running on different threads. If we're going that route it'll be necessary to check if there's any unsafe behavior downstream. In theory the state store stuff should be threadsafe already but is there anything else?...

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 16 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: I don't want to put this more complicated logic which is really only needed for tests into the heart of the the offer processing loop.  Added locks, but keeping polling on `awaitOffersProcessed`.

Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Yeah that's fair, :+1:

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 18 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Spoke offline, should be good.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 19]
```diff
@@ -63,8 +87,87 @@ public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, P
...
+
+    /**
+     * All offers must have been presented to resourceOffers() before calling this.  This call will block until all
+     * offers have been processed.
+     * @throws InterruptedException
+     */
+    @VisibleForTesting
+    public void awaitOffersProcessed() throws InterruptedException {
+        while (true) {
+            synchronized (offersInProgress) {
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: not inProgressLock like at L129?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java [hunk 20]
```diff
@@ -63,8 +87,87 @@ public void registered(SchedulerDriver driver, Protos.FrameworkID frameworkId, P
...
+                if (offersInProgress.isEmpty()) {
+                    LOGGER.info("All offers processed.");
+                    return;
+                }
+            }
+
+            LOGGER.warn("Offers to be processed {} is non empty, sleeping for 500ms ...", offersInProgress);
```
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: May as well keep this within the sync block

Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Want to release lock periodically to allow progress.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 19 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: my mistake

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 18 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: It looks to me that ultimately `TaskStatus` messages end up in the `update` call of `DeploymentStep` in all the `DefaultPlanCoordinator` case.  This updates only the status member variable which whose access is synchronized.  Will keep looking though.  Tests are not passing pretty broadly.  Trying to get a local repro to see what's happening.

**Comment by gabrielhartmann** [comment]
So far no repro on my own cluster:
1. Kafka: no repro
2. Cassandra: ...

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 18 above)
Comment (gabrielhartmann) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: Mystery solved.  Adding unrelated lock in status handling in the DeploymentStep was the problem.  Fixed separately by @nickbp.  After that fix all tests pass, want to take one last look @nickbp?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java (hunk 20 above)
Comment (nickbp) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/AbstractScheduler.java]: I mean, the logger message where it's printing offersInProgress should probably be within the lock protecting that. Feel free to defer to separate PR
